### PR TITLE
Updates the template mode canvas padding and adds a back link

### PR DESCRIPTION
**Pull request by youknowriad**
This is a small iteration towards a better block template mode in the post editor like shown here https://github.com/WordPress/gutenberg/issues/27849#issuecomment-800372073

This does a few small things:

 - Adds a padding around the template in template mode to better distinguish between the template and the default mode.
 - Adds a "back" button to move back to the regular mode.
 - Shows the title in the header and sidebar of the template mode.

This probably need some design  polish.

Capture d’écran 2021-04-09 à 11 37 49 AM

### CODE REVIEW

**Comment by youknowriad** [comment]
I think the next step after this PR is going to be improvements about the saving flow. It seems based on the mockups and also because it does feel right that we'd want to keep the saving flow similar between the default mode and the template mode, because we can just go back and forth between the modes, so probably just use the "multi-entity" saving flow in both cases if there are multiple entities to be saved.

**Comment by youknowriad** [comment]
Something we might want to start thinking about in terms of design is how to unify the "template" in the sidebar with the original template selector we have in the "page attributes" panel of the sidebar. These are ultimately the same things and we need to unify somehow.

File: packages/edit-post/src/components/header/template-title/index.js [hunk 1]
```diff
@@ -25,11 +25,18 @@ function TemplateTitle() {
 		return null;
 	}
 
+	let templateTitle = __( 'Default' );
+	if ( template && template?.title?.raw ) {
```
Comment (ntsekouras) [inline packages/edit-post/src/components/header/template-title/index.js]: I feel like making this inline with a triple `||` could be more straightforward, but nothing so important. If you keep it this way, loose the first `template` check.

File: packages/edit-post/src/components/sidebar/post-template/index.js [hunk 2]
```diff
@@ -59,18 +59,20 @@ function PostTemplate() {
 		return null;
 	}
 
+	const templateTitle = (
+		<>
+			{ !! template && template?.title?.raw }
```
Comment (ntsekouras) [inline packages/edit-post/src/components/sidebar/post-template/index.js:64]: same comment for a triple `||` applies here as well

File: packages/edit-post/src/components/visual-editor/style.scss [hunk 3]
```diff
@@ -23,6 +23,14 @@
 	@supports (position: sticky) {
 		flex-basis: 100%;
 	}
+
+	&.is-template-mode {
+		padding: $grid-unit-60;
```
Comment (ntsekouras) [inline packages/edit-post/src/components/visual-editor/style.scss:28]: This feels pretty big :)

**Review by ntsekouras — COMMENTED** [review]
Thanks for working on this Riad. In general this works as advertized and might just need some design input.

What this testing of the PR made me think about more, is the issues with saving flow in relation to the history (undo/redo). They are unrelated to the PR but they need more work.

**Comment by jameskoster** [comment]
> Something we might want to start thinking about in terms of design is how to unify the "template" in the sidebar with the original template selector we have in the "page attributes" panel of the sidebar.

Some explorations for this can be found in https://github.com/WordPress/gutenberg/issues/27850.

> so probably just use the "multi-entity" saving flow in both cases if there are multiple entities to be saved

Totally. On a somewhat related note, with the introduction of the <- Back button inside the frame, we can probably remove the "Cancel" button from the Top Bar. I think there's a lot of value in the Top Bar remaining consistent as one flips between the two editors which means the primary button should have the same label wherever possible, and the Preview menu should be added when editing a template. 

There is also the question of whether the "Template" tab from the site editor should replace the "Post" / "Page" tab in the Inspector accordingly.

I appreciate these changes will need to happen in follow-up PRs but figured they were worth noting none-the-less.

**Review by jameskoster — DISMISSED** [review]
In terms of actually merging this PR, I think it would be good to implement an interaction where clicking the frame deselects all blocks.

Edit: We also need to consider what to do when the background color is close to (or the same as) the frame color. Something like a subtle shadow, or a stroke can help:

Screenshot 2021-04-09 at 14 29 03

A more extreme option would be to allow users to manually select a color of their choosing.

**Comment by youknowriad** [comment]
It does feel like the frame is a bit too dark too (especially behind the back link) so maybe we should consider restoring the higher one @jasmussen

**Comment by jasmussen** [comment]
Agreed. I think we can tune things. I'd love a @mtias sanity check as well.

File: packages/edit-post/src/components/header/template-title/index.js (hunk 1 above)
Comment (youknowriad) [reply packages/edit-post/src/components/header/template-title/index.js]: we have three cases so it's not really possible  to inline or at least not with a single condition which means it becomes harder to read for me.

**Comment by youknowriad** [comment]
I've updated the PR, used a lighter background and added a light shadow as well. clicking the frame also deselects the block and I removed the "cancel" button. This is ready for another review.

File: packages/block-editor/src/components/block-selection-clearer/index.js [hunk 4]
```diff
@@ -9,7 +9,7 @@ import { useRefEffect } from '@wordpress/compose';
  */
 import { store as blockEditorStore } from '../../store';
 
-export function useBlockSelectionClearer() {
+export function useBlockSelectionClearer( onlySelfClicks = false ) {
```
Comment (ntsekouras) [inline packages/block-editor/src/components/block-selection-clearer/index.js:12]: Why this change was needed here?

**Comment by jameskoster** [comment]
Joen and I did a bunch of exploration around colors for the frame on Friday ([figma](https://www.figma.com/file/7u8wYpKOI4Zj9eXqComdEM/Conceptual-Frame?node-id=0%3A1)) and settled on the following:

* $gray-100 for the frame background
* 1px $gray-300 for the canvas border

With this configuration the edge of the canvas is always visible, even when the theme background matches the frame background or the canvas border. It also has a very small footprint in terms of introducing new UI since the border color matches other UI elements.

As an acid test I mocked this up using all the various designs we've shared in recent weeks that include the frame concept. I used the Eskell theme as an example, and tried an inverted version as well:

Screenshot 2021-04-12 at 12 12 41

You can take a closer look in the figma file [here](https://www.figma.com/file/7u8wYpKOI4Zj9eXqComdEM/Conceptual-Frame?node-id=10%3A70366).

This is by no means presented as a final design. But it feels like a solid foundation upon which we can build and iterate.

**Comment by jasmussen** [comment]
> This is by no means presented as a final design. But it feels like a solid foundation upon which we can build and iterate.

Yep, by "settled on" I'd like to clarify that we explored numerous options. White background, black background, drop shadows, various colors, all of them came with opportunities and tradeoffs, and none of them solved the problem in a blanket way. The black/dark background did look promising, especially when bleeding in to the top toolbar. But when the inspector was then opened, that needed to invert as well, just like the toolbar. Which is to say that's an exploration we can still do, but it's a much bigger mouthful.

Before proceeding, I'd love thoughts by @mtias as I think he's thought about this lately.

File: packages/block-editor/src/components/block-selection-clearer/index.js (hunk 4 above)
Comment (youknowriad) [reply packages/block-editor/src/components/block-selection-clearer/index.js:12]: Without it, clicking a button inside the frame would deselect the blocks and lose focus (like clicking an "appender" button) (A number of e2e tests were failing, that's how I discovered it)

**Comment by youknowriad** [comment]
I applied the suggested colors, the border colors feels a bit off to me though, I'd prefer a darker one or a shadow. This feels a bit blurry.

**Comment by youknowriad** [comment]
I guess, there's no perfect solution, the blurriness is visible in a theme like 2021 with the green bg but it's decent in a white bg theme.

**Comment by jameskoster** [comment]
For me this is good :) Let's see what Matias says.

**Comment by mtias** [comment]
I still want to try the darker background which I think offers a stronger separation, but we can do that separately. The placement and the wording of the "Back" button doesn't seem great to me.

One thing that jumps to mind, is that when switching to template editing, the inspector should switch from "page" to "template":

Current:
image

Ideally:
image

**Review by mtias — APPROVED** [review]

**Comment by mtias** [comment]
One other thing: this treatment should also be applied when editing a template part in isolation. (I still see the current no-canvas approach there.)

**Comment by jameskoster** [comment]
> One thing that jumps to mind, is that when switching to template editing, the inspector should switch from "page" to "template"

Yes this feels very important. I'll open an issue.

> One other thing: this treatment should also be applied when editing a template part in isolation.

This is one of many changes proposed in https://github.com/WordPress/gutenberg/issues/29148

**Comment by mtias** [comment]
One more issue: "select" mode seems to not be working as expected here (single click enters into edit mode sometimes).

**Comment by jameskoster** [comment]
That may be a broader site editing issue. There are few quirks with Select mode there (click-through pattern doesn't work at all).

**Comment by jasmussen** [comment]
🎉

**Comment by youknowriad** [comment]
Ok let's get this in. I'll be working on more follow-ups.

**Comment by mtias** [comment]
One thing I'd love to see us try is animating the background when entering this mode so that you can see the canvas going smaller. :)
